Implement smart contract functionality

Digital-Mercenaries opened the pull request:
I'm not too happy about onchain DevExp for `struct` when accessing from third-party contract, and implementing `ZorpStudy.claimReward` is something for my future self to figure out, but other than that most of the new features are ready for critique.

Digital-Mercenaries committed:
Use explicit imports to give IDEs an easier time

Digital-Mercenaries committed:
:tada: Implement most ZorpStudy functions!...

**Change highlights**

- integrate `ZorpStudy.constructor` changes within `createStudy`
- add unit tests for functions so far implemented
- add "TODO" comments for future self to do
- _bookmark_ ZorpStudy function that needs implemented with failing test

wkyleg commented on zorp-contracts/src/ZorpStudy.sol:
```diff
@@ -1,13 +1,46 @@
+contract ZorpStudy is Ownable, ReentrancyGuard {
+    uint256 public constant PARTICIPANT_STATUS__NA = 0;
+    uint256 public constant PARTICIPANT_STATUS__SUBMITTED = 1;
+    uint256 public constant PARTICIPANT_STATUS__INVALID = 2;
```
Why use `constant` over an Enum?

Digital-Mercenaries replied:
Partly out of habit, and convenience for third-party consumers both on and off the chain.  Because, if I remember correctly, both `enum` and `struct` are kinda annoying to deal with; [`ZorpTest.1.sol` line `110`][0] is an example of `struct` friction.

Partly for my convenience too; tab completion during development is faster when an editor need only match words instead of searching an AST ;-)

But I've a hunch it might be cheaper, in gas, to use `enums` instead when this is published as I think the compiler _should_ remove 'em.

[0]: https://github.com/Elata-Biosciences/zorp/pull/1/files#diff-c7ad0d5540e12ba99f60f843d4d126a4dfe6b14039b3dfae137708583ea3a458R110

wkyleg reviewed (approved):
Approving, just want to review design description of using constants over Enums. We can change this later need be

Digital-Mercenaries committed:
Implement `ZorpStudy.claimReward`...

Update `ZorpStudy.endStudy` to preform payout calculations and reclaim
any remainder caused by rounding errors to the study owner's account.

**Change highlights**

- `ZorpStudy.constructor` now requires `msg.value` to be positive
- `ZorpStudy.creator` points back to `ZorpFactory`
- functions now require addresses to be payable
- functions sending funds have reentry guards
- most functions are now `payable` for gas savings
- tests have been refactored to reduce some redundant repetition
- more `TODO` comments have been gifted to future self ;-)

Digital-Mercenaries commented:
`ffabd5b` is fresh and ready for review!  Main focus of my attentions, between last commits and now, were `ZorpStudy.claimReward()` and `ZorpStudy.endStudy()` functions.

Because Solidity doesn't do floating-point math and, last I checked it rounds down, I've encoded things such that study participants get an equal split with any rounding remainder going to the `ZorpStudy.owner()`.

Other than maybe converting `_STATUS__` constants to `enum`, and any high-priority to-dos which can be listed via;

```bash
grep -i -- todo src/*.sol test/ZorpTest.t.sol
```

...  I'm reasonably confident in implementations so far.  So unless I read otherwise I'll be adding some `interface` contracts that _should_ make consumption and doc-comments a bit nicer.

Digital-Mercenaries committed:
:dizzy: Re-order storage and functions, for gas...

And add `interface` contract for collecting doc-comments and hopefully
interactions between third parties a bit smoother too.

Digital-Mercenaries committed:
:fire: Remove `Participant` and add doc-comments

Now when `ZorpStudy.flagInvalidSubmission` is called data submitted by
participant will be deleted, check [gas.md#a7-sstore][] for breakdown
about gas refund implications.  Beyond that this change also _should_
help `ZorpStudy.owner()` when consuming valid submissions as empty
results from `submitted_data` can be ignored instead of incurring an
extra on-chain query for if associated account was flagged.

[gas.md#a7-sstore]: https://github.com/wolflo/evm-opcodes/blob/main/gas.md#a7-sstore

Digital-Mercenaries committed:
Add `ZorpStudy.paginateSubmittedData`...

This allows for off-chain callers to obtain multiple IPFS CID pointers
with on call, though there seems to be an issue in `forge` causing
explosive memory usage :shrug:

Adding JavaScript®/TypeScript™ based testing may be a wise choice in the
future, at least to provide usage example(s) for `ZorpStudy.owner()`

Digital-Mercenaries committed:
:dizzy: Move pagination to factory...

**Change highlights**

- add `IOwnable` interface to `ZorpStudy` and `ZorpFactory` interfaces
- add `IZorpFactory` interfaces
- move and add pagination to ZorpFactory to save the gas
- add pagination tests, though forge still no likes dynamic sized arrays
- contract storage slots now follow consistent naming conventions

wkyleg reviewed (approved):
LGTM, we can revisit refactoring with enums later

Digital-Mercenaries commented:
> LGTM, we can revisit refactoring with enums later

Thanks!

Indeed, I figure both refactoring, and doc-comments, can be done as we get the front-end side of things sketched-out more...  Speaking of which, ya can expect another `WiP` PR to _soon_ be submitted ;-)

Digital-Mercenaries commented:
> LGTM, we can revisit refactoring with enums later

So fun fact my past self's hunch about gas was not fictitious...  err at least it seems to be factual according to tests preformed on 2025-03-21...  `enum` **do** use more gas than `uint256 constant` for most operations within the `ZorpStudy` contract!

Here's some of the data collected via `forge test --gas-report`;

| Type                | Function Name                          | Min             | Avg   | Median | Max    | # Calls |
| ------------------- | -------------------------------------- | --------------- | ----- | ------ | ------ | ------- |
| `enum`              | `flagInvalidSubmission`                | 28991           | 37654 | 28991  | 54980  | 3       |
| `uint256 constant`  | `flagInvalidSubmission`                | 28795           | 37400 | 28795  | 54612  | 3       |
| `enum`              | `endStudy`                             | 31021           | 41109 | 42546  | 58411  | 5       |
| `uint256 constant`  | `endStudy`                             | 30956           | 40940 | 42309  | 58174  | 5       |
| `enum`              | `claimReward`                          | 28827           | 33521 | 31142  | 40595  | 3       |
| `uint256 constant`  | `claimReward`                          | 28762           | 33341 | 30990  | 40271  | 3       |

...  Note; I did my best to make it as close to fair as possible a comparison.  For example prior to switching over to `enum`, I added the explicit `uint` conversions that'll be required to make `assertEq` happy with future comparisons;

```diff
diff --git a/zorp-contracts/test/ZorpTest.t.sol b/zorp-contracts/test/ZorpTest.t.sol
index c99fc80..1cf2758 100644
--- a/zorp-contracts/test/ZorpTest.t.sol
+++ b/zorp-contracts/test/ZorpTest.t.sol
@@ -166,7 +166,7 @@ contract ZorpTest is Test {
         }

         assertEq(IZorpStudy(newStudy).invalidated(), 1, "Failed to increase invalidated count");
-        assertEq(IZorpStudy(newStudy).participant_status(ZORP_STUDY__PARTICIPANT__BAD), IZorpStudy(newStudy).PARTICIPANT_STATUS__INVALID(), "Failed to invalidate participant");
+        assertEq(IZorpStudy(newStudy).participant_status(ZORP_STUDY__PARTICIPANT__BAD), uint(IZorpStudy(newStudy).PARTICIPANT_STATUS__INVALID()), "Failed to invalidate participant");
     }

     function test_ZorpStudy_claimReward() public {
```

And checking the [Enums][0] docs says;

> "Enums cannot have more than 256 members."

... which may be a hint as to why there's a price difference.

[0]: https://docs.soliditylang.org/en/latest/types.html#enums
